Improve text editor status bar and zooming UX

MajorMcDoom opened the pull request:
https://github.com/godotengine/godot/assets/13228932/0ab33357-1239-4bf1-832e-e11cb9f9c3cd

**Old:**
![Screenshot 2024-02-17 214007](https://github.com/godotengine/godot/assets/13228932/21853a1d-cc2a-4455-b854-48b2672ef042)
**New:**
![Screenshot 2024-02-17 220131](https://github.com/godotengine/godot/assets/13228932/c86dbabe-7073-467a-9962-002f7fcbf5f0)


### Changes
- Text editor zooming no longer triggers a big editor freeze after completion. Fixes #82144.
- Text editor zooming feature now works by adding a multiplier on top of the editor code font size setting, instead of directly modifying the editor setting itself. This is consistent with most modern IDEs.
- The zoom level is now a _transient state_ of each text editor instance. This means for example that it can be different between the script editor and the shader editor, ~and even between each script tab~.
- The zoom level is now displayed in the status bar alongside the line-and-column stats. The zoom level is a clickable `MenuButton` that allows the user to quickly select between a few common zoom levels.
- The status bar now uses of a few `VSeparator`s to separate the various stats, instead of relying on vague empty space and a printed `|` character like before.
- The indentation indicator (`Tabs` / `Spaces`) has been separated from the line-and-column indicators and responds to the appropriate changes (e.g. indentation conversions and editor indentation setting changes), instead of responding to caret changes.
- Some cleanup of code to remove redundancies.

### Rationale
This represents a pretty big change to the role of the zoom feature in the user's workflow. Previously, the zoom feature was a shortcut for making a persistent change to the editor itself, even though it was not obvious to the user that this was happening. It was also easy to accidentally trigger.

In the new system, the zoom feature behaves as a quick, transient operation - similar to a scroll or a drag, while the editor code font size setting serves as the preferred neutral size (at 100% zoom). In this way, the purpose of the zoom feature is to give the user a quick way to navigate their _view_ of a document. ~They are free to leave some documents zoomed out for at-a-glance interaction, while zooming in on others where they are focused on specific areas.~

### Notes
- This is my first time contributing changes that involve the editor interface. Please excuse me if I did some weird things out of ignorance.
- The zoom levels in the popup are chosen arbitrarily based on my own intuition, and are free to change if there is a good reason to.
- ~The zoom levels of each open document are currently _not persistent,_ and will be reset between each session. However, they can totally be saved to the editor cache if we decide that is worthwhile (perhaps in another PR).~

MajorMcDoom committed:
Improved text editor status bar and zooming UX.

MewPurPur commented:
I prefer if all my scripts are at a persistent zoom level and it seems like this will make it hard to do so. Beside this, I fully agree with the approach of not changing a whole editor setting to achieve zooming.

MajorMcDoom commented:
> I prefer if all my scripts are at a persistent zoom level and it seems like this will make it hard to do so. Beside this, I fully agree with the approach of not changing a whole editor setting to achieve zooming.

@MewPurPur Just to clarify, do you mean _persistent_ (saves between sessions), or _consistent_ (synced between all tabs)?

Mickeon commented:
I myself am concerned about exposing this `indent_type_changed` signal. It's way too niche, I feel. Consider another way of accomplishing this without requiring the signal.

MajorMcDoom commented:
> I myself am concerned about exposing this `indent_type_changed` signal. It's way too niche, I feel. Consider another way of accomplishing this without requiring the signal.

@Mickeon I was wondering about that actually. How does one accomplish a subscription-like behaviour in the Godot codebase without the use of signals?

Mickeon commented:
There are some things in the API that are "exposed" but not really "exposed". These are prefixed with an underscore and do not show up on autocompletion or documentation, and by all means are discouraged to be used. I believe there's _never_ been a case of this for signals, however, so it's probably not supported at all.

So right now, I am not sure how. Again, I'm not against adding more things, but that would warrant people will actually need and use it.

If you want to ask for some help and feedback remember that Godot also has a public chat specifically tailored for development of the engine: https://chat.godotengine.org

Mickeon commented:
Right off the bat, I do have a suggestion. Thinking outside of the box.

A new signal would be much more beneficial if it was emitted on more than one situation, not just when _specifically_ the tab type changes. Resources have a `changed` signal. TileMap also has the `changed` signal as well ~~_(the doc's description is lying to you, what the hell, it is actually emitted in more situations)_~~.
There's many settings that influence **TextEdit**/**CodeEdit**'s behaviour. If they all emitted the same signal on change it would _probably_ be a whole lot more beneficial. 
Performance is not a concern, either, since these settings are unlikely to change very frequently, and the settings' previous values can be cached and checked before acting, anyhow.

MewPurPur commented:
It'd be better to first do the less controversial changes in this PR so it's approved more easily.

MajorMcDoom commented:
> It'd be better to first do the less controversial changes in this PR so it's approved more easily.

@MewPurPur  It's not quite so simple, since the "less controversial" fixes were only possible by decoupling from editor settings. And by doing so, all the script tabs now have to do additional work to stay in sync with each other or save their zoom. I thought it would be one interface, but no, every script in the tab actually holds onto its own text editor in memory.

That said, I am working on some changes.

MajorMcDoom commented:
@Mickeon After doing some more deep studying of the class hierarchy and their interplay, I implemented and pushed a solution which doesn't involve any changes to `CodeEdit`. Would this alleviate your concerns?

Mickeon commented:
It does, **marginally**. Now this feature can be taken a look at without having any interference with the public API, and it can be changed later if necessary.

I for one approve a lot but @MewPurPur made a good point before and it would be nice to see addressed. Never seen a "Per document" zooming level in an IDE (even if it does sound interesting).

MewPurPur commented:
My approach would've been to save one value in the editor state, just a script zoom level that'd hold across all scripts. This would be consistent with the current behavior, just no lag spikes from changed editor settings.

Calinou commented:
Script zoom should definitely be consistent across saves. Persistence is a bonus but not strictly required.

Web browsers store zoom levels on a per-domain basis, but this behavior isn't really suited to code editors.

MajorMcDoom commented:
@Calinou @MewPurPur @Mickeon 

I've added zoom consistency between different tabs of each editor. However, the Script Editor and the Shader Editor each have their own zoom levels, and I think that's a good thing.

I've also added persistence, implemented through the script and shader editor plugins themselves.

It was a huge pain to get access to the `CodeTextEditor` (where the zoom functionality is implemented) from the higher-level classes, especially across the two very differently-implemented script and shader systems. But I went for shortest path and added a couple of methods, and did some renamings to reduce confusion.

Mickeon commented:
Outstanding job. I hope this gets evaluated soon. The zooming was always a massive pain of the current text editor. It got to the point where @KoBeWi wanted to disable the shortcut outright for accidentally triggering it.

MajorMcDoom commented:
> It's not as smooth for me:
That's definitely the dev build. I've also experienced this on dev builds but it wasn't to your extreme. It was also present on dev builds for the old zoom before my changes. But let me know if this isn't the case on your end!

> Overall works great, I just found a small bug. When you set 75% zoom and restart, it will be restored as 79%.
This is actually intended behaviour. The zoom factor is always corrected to the nearest font size. It will not always land on a nice round number like the presets. This makes sure the zooming happens in consistent steps of 1 font size each. When you first select a zoom preset from the dropdown, it forces a label change to match the exact text you clicked, otherwise it looks like too much of a bug. But when you zoom with the shortcuts again, or restart, it will revert to displaying the actual ground truth zoom factor. This is also how Visual Studio behaves.

The alternative is that zooming happens in bigger increments and we don't correct the zoom factor to the nearest font size. This would result in bigger, more uneven zooming, with some font sizes being skipped, and potentially bigger accidental skips.

MajorMcDoom commented on editor/plugins/text_shader_editor.h:
```diff
@@ -148,7 +148,7 @@ class TextShaderEditor : public MarginContainer {
 	ConfirmationDialog *disk_changed = nullptr;
 
-	ShaderTextEditor *shader_editor = nullptr;
+	ShaderTextEditor *code_editor = nullptr;
```
This is the case which I said was to reduce confusion. The variable isn't exposed so I thought it would be fine to rename.

The containing class is TextShaderEditor, and is referred to by the plugin as shader_editor. This is fine, it's an editor of text shaders, so it's ultimately a shader editor. However the class also itself contains a variable called shader_editor. It also happens to be the variable I needed to expose. It would've resulted in shader_editor.get_shader_editor(). This variable's type is ShaderTextEditor, which is a subclass of CodeTextEditor. It's ultimately an editor of text. So a good renaming is to make it consistent with other classes on the script side of things, and call it a text_editor.

With this in mind, should I still move the variable rename to a different PR? I would like to at least keep the naming of the function I added (get_text_editor).

MajorMcDoom replied:
@KoBeWi I've fixed your other reviews. For this one, could you explain to me the rationale behind shying away from renaming a private variable like this? And if we absolutely have to keep the old name for the private var, is it okay if we have a `get_code_editor()` method that returns a var called `shader_editor`?

KoBeWi replied:
This change creates noise in the diff, but since there weren't other significant changes in that file, maybe it's fine in the end.

KoBeWi commented on editor/code_editor.h:
```diff
@@ -38,6 +38,7 @@
 #include "scene/gui/dialogs.h"
 #include "scene/gui/label.h"
 #include "scene/gui/line_edit.h"
+#include "scene/gui/menu_button.h"
```
This should be forward declared and included in cpp instead.
(same applies to most of the includes here, but they are out of scope of this PR)

KoBeWi commented on editor/code_editor.h:
```diff
@@ -234,6 +227,8 @@ class CodeTextEditor : public VBoxContainer {
 	};
 	void convert_case(CaseStyle p_case);
 
+	void set_indent_using_spaces(const bool p_use_spaces);
```
```suggestion
	void set_indent_using_spaces(bool p_use_spaces);
```
Primitive parameters don't need `const`.

MajorMcDoom replied:
No idea how that got there. Will fix this plus the cpp.

MajorMcDoom commented:
> > This is actually intended behaviour. The zoom factor is always corrected to the nearest font size.
> 
> This makes sense but it definitely does not look correct. I can imagine an unknowing user being ever so slightly bothered by this to open an issue. Not sure what the exact solution could be. Maybe the display is rounded down to the specific percentage (e.g. `79%` -> `75%`)? Alternatively the way the zoom is displayed could be re-evaluated? Visual Studio Code displays an lens icon if the whole editor is zoomed in or out. But that's not something done anywhere in Godot, where the percentage is much more common.

Rounding isn't an option because there is no rounding interval which is simultaneously:
- big enough to mask cases like this one.
- small enough to always show a change when you zoom (lack of UI feedback)

I'm a little unclear as to what your second alternative was, could you explain that one again?

One solution I can propose is to bypass the scale factor correction itself when setting to one of the presets, instead of just masking with a display change. I will try that.

KoBeWi commented on editor/code_editor.cpp:
```diff
@@ -1785,6 +1753,13 @@ void CodeTextEditor::_warning_button_pressed() {
 
+void CodeTextEditor::_zoom_popup_id_pressed(int p_idx) {
+	const String &text = zoom_button->get_popup()->get_item_text(p_idx);
+	float new_zoom_factor = text.trim_suffix(" %").to_float() / 100.0f;
```
A cleaner solution than parsing text would be setting item metadata.

KoBeWi commented on editor/code_editor.cpp:
```diff
@@ -2087,13 +2044,42 @@ CodeTextEditor::CodeTextEditor() {
+	zoom_button->set_v_size_flags(SIZE_EXPAND | SIZE_SHRINK_CENTER);
+	zoom_button->set_text("100 %");
+	zoom_button->set_tooltip_text(TTR("Zoom factor"));
+	zoom_button->get_popup()->add_item("25 %");
```
Should this space be here? (the one between number and `%`, it looks odd).

Mickeon replied:
I'm not sure but it seems the editor has a space, too. (Just not on the tooltip). 
![image](https://github.com/godotengine/godot/assets/66727710/471fc029-3ebb-4a14-9e6f-24b7955fcace)
If yes, keeping it would be consistent.

MajorMcDoom replied:
Ah yeah that was also another thing I checked against for consistency. I thought it was odd at first but after trying it without the space, the menu looked like a contiguous blob of various sized circles due to all the rows of 00%. The default editor font also makes this worse because the percent sign is very big and prominent.

KoBeWi commented on editor/code_editor.cpp:
```diff
@@ -1956,11 +1908,47 @@ void CodeTextEditor::remove_all_bookmarks() {
+
+	if (new_zoom_factor != zoom_factor) {
+		zoom_factor = new_zoom_factor;
+		emit_signal(SNAME("zoomed"), new_zoom_factor);
```
The signal causes `set_zoom_factor()` to be called in every CodeTextEditor, which emits the signal again and updates every script editor. So with e.g. 4 script tabs, you get 16 update calls. This should be sorted out, so that zoom change only updates every editor once.

MajorMcDoom replied:
The shouldn't happen. There are multiple layers of redundancy checks to prevent this.

One tab detects interaction, sets new value, fires signal. That's 1 update. The plugin catches signal, then propagates the zoom to only the tabs that have different zoom. For 4 tabs, that's 3 updates. No additional updates will happen after this now that the zooms are all equal.

KoBeWi replied:
Right, but the `ScriptEditor::_set_zoom_factor` is still called after every emission (which happens in every code editor when you change zoom) and it calls methods in every code editor. Even if they stop at the zoom comparison, that's still many unnecessary calls.

KoBeWi replied:
> The only redundancy I can see here is that each time ScriptEditor catches a signal, it goes through a loop to check all the tabs. Is that what you mean by unnecessary calls?

Yes.

KoBeWi commented on editor/plugins/script_editor_plugin.h:
```diff
@@ -32,6 +32,7 @@
 #define SCRIPT_EDITOR_PLUGIN_H
 
 #include "core/object/script_language.h"
+#include "editor/code_editor.h"
```
This should also be forward declared.

MajorMcDoom commented on editor/code_editor.cpp:
```diff
@@ -2087,13 +2044,42 @@ CodeTextEditor::CodeTextEditor() {
+	status_bar->add_child(memnew(VSeparator));
+
+	// Zoom
+	zoom_button = memnew(MenuButton);
```
I remembered the other reason I chose a MenuButton is because the text isn't always a reflection of the selected text, and in fact, there isn't even always a selection. So constantly managing the text override and the selection index when there's a custom zoom (which is most of the time), was less straightforward than just using a MenuButton, which is what a lot of things in the editor's bars use.

Unless there's another good reason not to use a MenuButton I'm gonna mark this one as resolved.

MajorMcDoom commented:
@Mickeon @KoBeWi 
I've changed `CodeTextEditor::set_zoom_factor`, and now the presets will now display properly when closing and restarting the editor. Thanks for catching that.

Calinou commented on editor/code_editor.cpp:
```diff
@@ -1956,11 +1908,49 @@ void CodeTextEditor::remove_all_bookmarks() {
+
+	int preset_count = sizeof(ZOOM_FACTOR_PRESETS) / sizeof(float);
+	float new_zoom_factor = CLAMP(p_zoom_factor, ZOOM_FACTOR_PRESETS[0], ZOOM_FACTOR_PRESETS[preset_count - 1]);
+	int neutral_font_size = EDITOR_GET("interface/editor/code_font_size").operator int() * EDSCALE;
```
Use simpler casting form:

```suggestion
	int neutral_font_size = int(EDITOR_GET("interface/editor/code_font_size")) EDSCALE;
```

Calinou commented on editor/code_editor.cpp:
```diff
@@ -1956,11 +1908,49 @@ void CodeTextEditor::remove_all_bookmarks() {
+	int preset_count = sizeof(ZOOM_FACTOR_PRESETS) / sizeof(float);
+	float new_zoom_factor = CLAMP(p_zoom_factor, ZOOM_FACTOR_PRESETS[0], ZOOM_FACTOR_PRESETS[preset_count - 1]);
+	int neutral_font_size = EDITOR_GET("interface/editor/code_font_size").operator int() * EDSCALE;
+	int new_font_size = round(new_zoom_factor * neutral_font_size);
```
I noticed this uses `round()` instead of `Math::round()`. It's not a dealbreaker for this PR but this struck me as odd. Some other places in the engine already use `round()`, such as https://github.com/godotengine/godot/blob/652438a3952ae13e12dd4fd40756311fde81a791/servers/audio/effects/eq_filter.cpp#L89.

@godotengine/core Any insights on this? We should probably do a pass throughout the codebase to harmonize on this.

Calinou commented on editor/code_editor.cpp:
```diff
@@ -757,6 +759,8 @@ FindReplaceBar::FindReplaceBar() {
 
 /*** CODE EDITOR ****/
 
+const float CodeTextEditor::ZOOM_FACTOR_PRESETS[7] = { 0.25f, 0.5f, 0.75f, 1.0f, 1.5f, 2.0f, 3.0f };
```
I think you can add `static` before `const` here so you don't need to declare it in the header.

Calinou commented on editor/code_editor.cpp:
```diff
@@ -757,6 +759,8 @@ FindReplaceBar::FindReplaceBar() {
 
 /*** CODE EDITOR ****/
 
+const float CodeTextEditor::ZOOM_FACTOR_PRESETS[7] = { 0.25f, 0.5f, 0.75f, 1.0f, 1.5f, 2.0f, 3.0f };
```
This should allow declaring the variable only here and removing the declaration in the header:

```suggestion
static constexpr float CodeTextEditor::ZOOM_FACTOR_PRESETS[7] = { 0.25f, 0.5f, 0.75f, 1.0f, 1.5f, 2.0f, 3.0f };
```

MajorMcDoom replied:
Did you mean to take out the `CodeTextEditor::` by any chance?

Calinou replied:
It can probably be taken out too.

MajorMcDoom replied:
It actually can't compile otherwise. 😅

Calinou reviewed (approved):
Tested locally (at 100% and 175% editor scales), it works as expected.

Note that the first time you zoom in an editor session (causing a new font size to be displayed), there will be some slowdown due to the font cache having to be regenerated. This doesn't occur afterwards. Using MSDF rendering for the code font would alleviate this, at the cost of making it less crisp (so I don't recommend this approach, at least for now).

KoBeWi commented on editor/plugins/shader_editor_plugin.cpp:
```diff
@@ -590,6 +601,22 @@ void ShaderEditorPlugin::_window_changed(bool p_visible) {
+		for (const EditedShader &edited_shader : edited_shaders) {
+			if (edited_shader.shader_editor) {
+				CodeTextEditor *cte = edited_shader.shader_editor->get_code_editor();
+				if (cte) {
+					if (cte->get_zoom_factor() != text_shader_zoom_factor) {
```
```suggestion
				if (cte && cte->get_zoom_factor() != text_shader_zoom_factor) {
```
Less nesting.

KoBeWi replied:
You could also apply this in ScriptEditor, as it has a similar part.

KoBeWi commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -3801,6 +3811,23 @@ void ScriptEditor::_on_find_in_files_modified_files(PackedStringArray paths) {
 }
 
+void ScriptEditor::_set_zoom_factor(float p_zoom_factor) {
+	if (zoom_factor != p_zoom_factor) {
```
This can be early return instead:
```suggestion
	if (zoom_factor == p_zoom_factor) {
		return
	}
```

KoBeWi reviewed (approved):
Looks fine now.

Needs rebase after #69032
Note that every function that updates editor settings (either from `NOTIFICATION_EDITOR_SETTINGS_CHANGED` or from `settings_changed` signal) should use `check_changed_settings_in_group()` now.

MajorMcDoom commented:
@KoBeWi Did the rebase! Should be ready to go. 🤞

akien-mga commented:
Thanks!
